File: hyperliquid-trendlines/hyperliquid_trendlines/trendlines.py
"""Trendline detection tuned to the breakout playbook from the research notes:

- Draw ascending lines along swing lows, descending lines along swing highs.
- Require >=3 touches to treat a line as valid.
- Reject lines that violate the price series (line crosses into candles).
- Prefer shallower slopes; flag lines steeper than ~60deg in log-price space.
- A breakout is a decisive close through the line, ideally with volume >=
  ~1.3x the trailing average.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Literal, Sequence

import math
import numpy as np
import pandas as pd


Kind = Literal["support", "resistance"]


@dataclass
class Trendline:
    kind: Kind
    slope: float                 # price per bar
    intercept: float             # price at bar index 0
    start_idx: int               # first pivot index
    end_idx: int                 # last pivot index used to fit
    pivots: List[int] = field(default_factory=list)
    touches: int = 0
    r_squared: float = 0.0

    def price_at(self, idx: int | np.ndarray) -> float | np.ndarray:
        return self.slope * np.asarray(idx) + self.intercept

    def angle_degrees(self, bar_span_price: float) -> float:
        """Angle in degrees after normalising bars and price to a common scale.
        ``bar_span_price`` is the typical price change per bar (e.g. ATR) used
        to make the angle meaningful instead of axis-dependent."""
        if bar_span_price <= 0:
            return 0.0
        return math.degrees(math.atan(self.slope / bar_span_price))


@dataclass
class Breakout:
    idx: int
    timestamp: pd.Timestamp
    price: float
    line: Trendline
    direction: Literal["up", "down"]
    volume_ratio: float
    retest: bool = False


def find_pivots(
    highs: Sequence[float],
    lows: Sequence[float],
    left: int = 3,
    right: int = 3,
) -> tuple[list[int], list[int]]:
    """Classical fractal pivots: a bar is a pivot high if its high is the max
    over [i-left, i+right], and similarly for pivot lows."""
    highs = np.asarray(highs, dtype=float)
    lows = np.asarray(lows, dtype=float)
    n = len(highs)
    pivot_highs: list[int] = []
    pivot_lows: list[int] = []
    for i in range(left, n - right):
        window_h = highs[i - left : i + right + 1]
        window_l = lows[i - left : i + right + 1]
        if highs[i] == window_h.max() and (window_h == highs[i]).sum() == 1:
            pivot_highs.append(i)
        if lows[i] == window_l.min() and (window_l == lows[i]).sum() == 1:
            pivot_lows.append(i)
    return pivot_highs, pivot_lows


def _fit_line(xs: np.ndarray, ys: np.ndarray) -> tuple[float, float, float]:
    """Least-squares line. Returns slope, intercept, r^2."""
    if len(xs) < 2:
        return 0.0, float(ys[0]) if len(ys) else 0.0, 0.0
    slope, intercept = np.polyfit(xs, ys, 1)
    preds = slope * xs + intercept
    ss_res = float(np.sum((ys - preds) ** 2))
    ss_tot = float(np.sum((ys - ys.mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot if ss_tot > 0 else 1.0
    return float(slope), float(intercept), r2


def _validate_line(
    slope: float,
    intercept: float,
    series: np.ndarray,
    start: int,
    end: int,
    kind: Kind,
    tol_pct: float = 0.001,
) -> tuple[bool, int]:
    """A support line must sit at-or-below lows; a resistance line at-or-above
    highs, between ``start`` and ``end``. Small tolerance allows wick noise.
    Returns (valid, touch_count) where a touch means the bar came within
    ``tol_pct`` of the line."""
    idxs = np.arange(start, end + 1)
    line_vals = slope * idxs + intercept
    window = series[start : end + 1]
    if kind == "support":
        tolerance = np.abs(line_vals) * tol_pct
        if np.any(window < line_vals - tolerance):
            return False, 0
        touches = int(np.sum(np.abs(window - line_vals) <= tolerance))
    else:
        tolerance = np.abs(line_vals) * tol_pct
        if np.any(window > line_vals + tolerance):
            return False, 0
        touches = int(np.sum(np.abs(window - line_vals) <= tolerance))
    return True, touches


def _best_line_through_pivots(
    pivots: Sequence[int],
    series: np.ndarray,
    kind: Kind,
    min_touches: int = 3,
    tol_pct: float = 0.001,
) -> Trendline | None:
    """Search pairs of pivots, extend the line across the whole window, keep
    the line with the most touches that doesn't violate the price series."""
    pivots = list(pivots)
    if len(pivots) < 2:
        return None

    n = len(series)
    best: Trendline | None = None

    for i in range(len(pivots)):
        for j in range(i + 1, len(pivots)):
            p1, p2 = pivots[i], pivots[j]
            x1, x2 = float(p1), float(p2)
            y1, y2 = float(series[p1]), float(series[p2])
            if x2 == x1:
                continue
            slope = (y2 - y1) / (x2 - x1)
            intercept = y1 - slope * x1

            valid, touches = _validate_line(
                slope, intercept, series, start=p1, end=n - 1,
                kind=kind, tol_pct=tol_pct,
            )
            if not valid or touches < min_touches:
                continue

            involved = [
                p for p in pivots
                if p1 <= p <= n - 1
                and abs(series[p] - (slope * p + intercept)) <= abs(slope * p + intercept) * tol_pct
            ]
            if len(involved) < 2:
                continue

            xs = np.array(involved, dtype=float)
            ys = series[involved]
            fit_slope, fit_intercept, r2 = _fit_line(xs, ys)

            valid2, touches2 = _validate_line(
                fit_slope, fit_intercept, series, start=p1, end=n - 1,
                kind=kind, tol_pct=tol_pct,
            )
            if valid2 and touches2 >= touches:
                slope, intercept, touches = fit_slope, fit_intercept, touches2

            candidate = Trendline(
                kind=kind, slope=slope, intercept=intercept,
                start_idx=min(involved), end_idx=max(involved),
                pivots=sorted(involved), touches=touches, r_squared=r2,
            )
            if best is None or (
                candidate.touches, candidate.r_squared
            ) > (best.touches, best.r_squared):
                best = candidate
    return best


def detect_trendlines(
    df: pd.DataFrame,
    left: int = 3,
    right: int = 3,
    min_touches: int = 3,
    tol_pct: float = 0.0015,
    max_lines_per_side: int = 2,
) -> list[Trendline]:
    """Find the most meaningful support and resistance trendlines on ``df``."""
    highs = df["High"].to_numpy()
    lows = df["Low"].to_numpy()
    pivot_highs, pivot_lows = find_pivots(highs, lows, left=left, right=right)

    lines: list[Trendline] = []

    # Resistance lines along pivot highs.
    res = _best_line_through_pivots(
        pivot_highs, highs, kind="resistance",
        min_touches=min_touches, tol_pct=tol_pct,
    )
    if res is not None:
        lines.append(res)

    # Support lines along pivot lows.
    sup = _best_line_through_pivots(
        pivot_lows, lows, kind="support",
        min_touches=min_touches, tol_pct=tol_pct,
    )
    if sup is not None:
        lines.append(sup)

    # Secondary lines: retry ignoring the pivots already used.
    if max_lines_per_side > 1 and res is not None:
        remaining = [p for p in pivot_highs if p not in res.pivots]
        extra = _best_line_through_pivots(
            remaining, highs, kind="resistance",
            min_touches=min_touches, tol_pct=tol_pct,
        )
        if extra is not None:
            lines.append(extra)

    if max_lines_per_side > 1 and sup is not None:
        remaining = [p for p in pivot_lows if p not in sup.pivots]
        extra = _best_line_through_pivots(
            remaining, lows, kind="support",
            min_touches=min_touches, tol_pct=tol_pct,
        )
        if extra is not None:
            lines.append(extra)

    return lines


def detect_breakouts(
    df: pd.DataFrame,
    lines: Sequence[Trendline],
    volume_window: int = 20,
    volume_multiplier: float = 1.3,
    min_break_pct: float = 0.001,
) -> list[Breakout]:
    """Mark the first decisive close through each line with volume support."""
    closes = df["Close"].to_numpy()
    volumes = df["Volume"].to_numpy()
    avg_vol = pd.Series(volumes).rolling(volume_window, min_periods=1).mean().to_numpy()
    n = len(df)

    breakouts: list[Breakout] = []
    for line in lines:
        for i in range(max(line.end_idx + 1, 1), n):
            level = line.price_at(i)
            if level <= 0 or not np.isfinite(level):
                continue
            close = closes[i]
            vol_ratio = volumes[i] / avg_vol[i] if avg_vol[i] > 0 else 0.0
            breached = abs(close - level) / level >= min_break_pct

            if line.kind == "resistance" and close > level and breached:
                direction = "up"
            elif line.kind == "support" and close < level and breached:
                direction = "down"
            else:
                continue

            retest = False
            for j in range(i + 1, min(i + 10, n)):
                lj = line.price_at(j)
                if direction == "up" and df["Low"].iloc[j] <= lj <= df["High"].iloc[j] and closes[j] > lj:
                    retest = True
                    break
                if direction == "down" and df["Low"].iloc[j] <= lj <= df["High"].iloc[j] and closes[j] < lj:
                    retest = True
                    break

            breakouts.append(
                Breakout(
                    idx=i,
                    timestamp=df.index[i],
                    price=float(close),
                    line=line,
                    direction=direction,
                    volume_ratio=float(vol_ratio),
                    retest=retest,
                )
            )
            break

    breakouts = [b for b in breakouts if b.volume_ratio >= volume_multiplier or b.retest]
    return breakouts

File: hyperliquid-trendlines-web/lib/trendlines.ts
import type { Candle } from "./hyperliquid";

export type Kind = "support" | "resistance";

export interface Trendline {
  kind: Kind;
  slope: number;       // price per bar (index units)
  intercept: number;   // price at bar 0
  startIdx: number;    // first pivot index used
  endIdx: number;      // last pivot index used
  pivots: number[];
  touches: number;
  rSquared: number;
}

export interface Breakout {
  idx: number;
  time: number;        // unix seconds
  price: number;
  direction: "up" | "down";
  volumeRatio: number;
  retest: boolean;
  line: Trendline;
}

export function priceAt(line: Trendline, idx: number): number {
  return line.slope * idx + line.intercept;
}

export function findPivots(
  highs: number[],
  lows: number[],
  left = 3,
  right = 3,
): { pivotHighs: number[]; pivotLows: number[] } {
  const pivotHighs: number[] = [];
  const pivotLows: number[] = [];
  const n = highs.length;
  for (let i = left; i < n - right; i++) {
    let isHigh = true;
    let isLow = true;
    let highMatches = 0;
    let lowMatches = 0;
    for (let k = i - left; k <= i + right; k++) {
      if (highs[k] > highs[i]) isHigh = false;
      if (lows[k] < lows[i]) isLow = false;
      if (highs[k] === highs[i]) highMatches++;
      if (lows[k] === lows[i]) lowMatches++;
    }
    if (isHigh && highMatches === 1) pivotHighs.push(i);
    if (isLow && lowMatches === 1) pivotLows.push(i);
  }
  return { pivotHighs, pivotLows };
}

function fitLine(xs: number[], ys: number[]): { slope: number; intercept: number; r2: number } {
  const n = xs.length;
  if (n < 2) return { slope: 0, intercept: ys[0] ?? 0, r2: 0 };
  let sumX = 0, sumY = 0, sumXY = 0, sumXX = 0;
  for (let i = 0; i < n; i++) {
    sumX += xs[i]; sumY += ys[i];
    sumXY += xs[i] * ys[i]; sumXX += xs[i] * xs[i];
  }
  const denom = n * sumXX - sumX * sumX;
  const slope = denom === 0 ? 0 : (n * sumXY - sumX * sumY) / denom;
  const intercept = (sumY - slope * sumX) / n;
  const meanY = sumY / n;
  let ssRes = 0, ssTot = 0;
  for (let i = 0; i < n; i++) {
    const pred = slope * xs[i] + intercept;
    ssRes += (ys[i] - pred) ** 2;
    ssTot += (ys[i] - meanY) ** 2;
  }
  const r2 = ssTot > 0 ? 1 - ssRes / ssTot : 1;
  return { slope, intercept, r2 };
}

function validateLine(
  slope: number,
  intercept: number,
  series: number[],
  start: number,
  end: number,
  kind: Kind,
  tolPct: number,
): { valid: boolean; touches: number } {
  let touches = 0;
  for (let i = start; i <= end; i++) {
    const lv = slope * i + intercept;
    const tol = Math.abs(lv) * tolPct;
    if (kind === "support") {
      if (series[i] < lv - tol) return { valid: false, touches: 0 };
    } else {
      if (series[i] > lv + tol) return { valid: false, touches: 0 };
    }
    if (Math.abs(series[i] - lv) <= tol) touches++;
  }
  return { valid: true, touches };
}

function bestLineThroughPivots(
  pivots: number[],
  series: number[],
  kind: Kind,
  minTouches: number,
  tolPct: number,
): Trendline | null {
  if (pivots.length < 2) return null;
  const n = series.length;
  let best: Trendline | null = null;

  for (let i = 0; i < pivots.length; i++) {
    for (let j = i + 1; j < pivots.length; j++) {
      const p1 = pivots[i];
      const p2 = pivots[j];
      const dx = p2 - p1;
      if (dx === 0) continue;
      let slope = (series[p2] - series[p1]) / dx;
      let intercept = series[p1] - slope * p1;

      let check = validateLine(slope, intercept, series, p1, n - 1, kind, tolPct);
      if (!check.valid || check.touches < minTouches) continue;

      const involved: number[] = [];
      for (const p of pivots) {
        if (p < p1 || p > n - 1) continue;
        const lv = slope * p + intercept;
        if (Math.abs(series[p] - lv) <= Math.abs(lv) * tolPct) involved.push(p);
      }
      if (involved.length < 2) continue;

      const xs = involved.map((x) => x);
      const ys = involved.map((p) => series[p]);
      const fit = fitLine(xs, ys);
      const refit = validateLine(fit.slope, fit.intercept, series, p1, n - 1, kind, tolPct);
      if (refit.valid && refit.touches >= check.touches) {
        slope = fit.slope;
        intercept = fit.intercept;
        check = refit;
      }

      const candidate: Trendline = {
        kind,
        slope,
        intercept,
        startIdx: Math.min(...involved),
        endIdx: Math.max(...involved),
        pivots: involved.slice().sort((a, b) => a - b),
        touches: check.touches,
        rSquared: fit.r2,
      };
      if (
        best === null ||
        candidate.touches > best.touches ||
        (candidate.touches === best.touches && candidate.rSquared > best.rSquared)
      ) {
        best = candidate;
      }
    }
  }
  return best;
}

export interface DetectOptions {
  left?: number;
  right?: number;
  minTouches?: number;
  tolPct?: number;
  maxLinesPerSide?: number;
}

export function detectTrendlines(candles: Candle[], opts: DetectOptions = {}): Trendline[] {
  const left = opts.left ?? 3;
  const right = opts.right ?? 3;
  const minTouches = opts.minTouches ?? 3;
  const tolPct = opts.tolPct ?? 0.0015;
  const maxLinesPerSide = opts.maxLinesPerSide ?? 2;

  const highs = candles.map((c) => c.high);
  const lows = candles.map((c) => c.low);
  const { pivotHighs, pivotLows } = findPivots(highs, lows, left, right);

  const lines: Trendline[] = [];

  const res = bestLineThroughPivots(pivotHighs, highs, "resistance", minTouches, tolPct);
  if (res) lines.push(res);
  const sup = bestLineThroughPivots(pivotLows, lows, "support", minTouches, tolPct);
  if (sup) lines.push(sup);

  if (maxLinesPerSide > 1 && res) {
    const rem = pivotHighs.filter((p) => !res.pivots.includes(p));
    const extra = bestLineThroughPivots(rem, highs, "resistance", minTouches, tolPct);
    if (extra) lines.push(extra);
  }
  if (maxLinesPerSide > 1 && sup) {
    const rem = pivotLows.filter((p) => !sup.pivots.includes(p));
    const extra = bestLineThroughPivots(rem, lows, "support", minTouches, tolPct);
    if (extra) lines.push(extra);
  }

  return lines;
}

export interface BreakoutOptions {
  volumeWindow?: number;
  volumeMultiplier?: number;
  minBreakPct?: number;
}

export function detectBreakouts(
  candles: Candle[],
  lines: Trendline[],
  opts: BreakoutOptions = {},
): Breakout[] {
  const volWindow = opts.volumeWindow ?? 20;
  const volMult = opts.volumeMultiplier ?? 1.3;
  const minPct = opts.minBreakPct ?? 0.001;

  const n = candles.length;
  const rollingAvg = new Array<number>(n).fill(0);
  {
    let sum = 0;
    const q: number[] = [];
    for (let i = 0; i < n; i++) {
      q.push(candles[i].volume);
      sum += candles[i].volume;
      if (q.length > volWindow) sum -= q.shift()!;
      rollingAvg[i] = sum / q.length;
    }
  }

  const breakouts: Breakout[] = [];
  for (const line of lines) {
    for (let i = Math.max(line.endIdx + 1, 1); i < n; i++) {
      const lv = priceAt(line, i);
      if (!isFinite(lv) || lv <= 0) continue;
      const c = candles[i];
      const deviation = Math.abs(c.close - lv) / lv;
      if (deviation < minPct) continue;

      let direction: "up" | "down" | null = null;
      if (line.kind === "resistance" && c.close > lv) direction = "up";
      else if (line.kind === "support" && c.close < lv) direction = "down";
      if (!direction) continue;

      let retest = false;
      for (let j = i + 1; j < Math.min(i + 10, n); j++) {
        const lj = priceAt(line, j);
        const cj = candles[j];
        if (cj.low <= lj && lj <= cj.high) {
          if (direction === "up" && cj.close > lj) { retest = true; break; }
          if (direction === "down" && cj.close < lj) { retest = true; break; }
        }
      }

      const volRatio = rollingAvg[i] > 0 ? c.volume / rollingAvg[i] : 0;
      if (volRatio >= volMult || retest) {
        breakouts.push({
          idx: i,
          time: c.time,
          price: c.close,
          direction,
          volumeRatio: volRatio,
          retest,
          line,
        });
      }
      break;
    }
  }
  return breakouts;
}
